File: src/integrations/replit.py
"""
Replit integration module.
Triggers Replit deployments and interacts with Replit projects via its API.
"""

import os
import httpx


class ReplitIntegration:
    """Handles Replit project deployment and status checks."""

    BASE_URL = "https://replit.com/graphql"

    def __init__(self):
        self.api_key = os.environ.get("REPLIT_API_KEY")
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "X-Requested-With": "XMLHttpRequest",
        }

    async def get_repl(self, repl_id: str) -> dict:
        """Fetch metadata for a specific Repl by ID."""
        if not self.api_key:
            return {"error": "REPLIT_API_KEY not configured"}

        query = """
        query GetRepl($id: String!) {
          repl(id: $id) {
            id
            title
            slug
            url
            isPrivate
            deploymentState
          }
        }
        """
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                self.BASE_URL,
                json={"query": query, "variables": {"id": repl_id}},
                headers=self.headers,
            )
            resp.raise_for_status()
            return resp.json()

    async def run_repl(self, repl_id: str) -> dict:
        """Trigger a run/deploy for a Repl."""
        if not self.api_key:
            return {"error": "REPLIT_API_KEY not configured"}

        mutation = """
        mutation RunRepl($replId: String!) {
          runRepl(input: { replId: $replId }) {
            repl {
              id
              title
              deploymentState
            }
          }
        }
        """
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                self.BASE_URL,
                json={"query": mutation, "variables": {"replId": repl_id}},
                headers=self.headers,
            )
            resp.raise_for_status()
            return resp.json()

File: src/connectors/replit.js
'use strict';

const axios = require('axios');
const logger = require('../logger');

function getClient() {
  const apiKey = process.env.REPLIT_API_KEY;
  const baseURL = process.env.REPLIT_BASE_URL || 'https://replit.com/api/v1';
  if (!apiKey) {throw new Error('REPLIT_API_KEY is not configured');}
  return axios.create({
    baseURL,
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    timeout: 30_000,
  });
}

/**
 * List Repls owned by the authenticated user.
 * @returns {Promise<Array>}
 */
async function listRepls() {
  logger.debug('Replit: listRepls');
  const client = getClient();
  const response = await client.get('/repls');
  return response.data;
}

/**
 * Create a new Repl.
 * @param {object} config  { title, language, description, isPrivate }
 * @returns {Promise<object>}
 */
async function createRepl(config) {
  logger.debug('Replit: createRepl', { title: config.title, language: config.language });
  const client = getClient();
  const response = await client.post('/repls', config);
  return response.data;
}

/**
 * Run (execute) a Repl by its ID.
 * @param {string} replId
 * @returns {Promise<object>}
 */
async function runRepl(replId) {
  logger.debug('Replit: runRepl', { replId });
  const client = getClient();
  const response = await client.post(`/repls/${replId}/run`);
  return response.data;
}

module.exports = { listRepls, createRepl, runRepl };
